### ISSUE TITLE
Trippy BSOD on NetBSD when resizing window

### ISSUE THREAD

**Issue by fujiapple852** [fujiapple852/trippy#276]
Starting trippy (`0.6.0-dev`) and resizing the window will result in a BSOD with the error:

```
IO error: Interrupted system call (os error 4)
```

**Comment by fujiapple852** [comment]
Also occurs with the existing `0.5.0` release and so not a new regression.  @0323pin do you recall if you ever saw this issue?

**Comment by 0323pin** [comment]
@fujiapple852 I don't because, I do not resize windows, I use `leftwm` (tiling window manager).

But, I can reproduce the error with `0.5.0` when **I do** a window resize.

**Comment by fujiapple852** [comment]
Thank you @0323pin!  Given it isn't a regression bug I'll proceed with the `0.6.0` release and investigate this with a view to fixing in `0.7.0`.

**Comment by 0323pin** [comment]
No worries, @fujiapple852 
Weird that this happens, though.

**Comment by 0323pin** [comment]
@fujiapple852 Just so you know, I've updated the package   already but, didn't have the time to merge it. Most probably late this evening or, tomorrow early morning.

**Comment by fujiapple852** [comment]
Added #552 to assist with diagnostic of this issue

**Comment by fujiapple852** [comment]
Best guess is this relates to not handling `EINT` properly. 

https://unix.stackexchange.com/questions/509375/what-is-interrupted-system-call

**Comment by 0323pin** [comment]
Sorry for the slow reply, been AFK for a few days.

I've asked internally, if we are missing something obvious here.

**Comment by fujiapple852** [comment]
Thanks @0323pin, there is nothing needed right now, I've managed to get an AWS NetBSD instance working again and so I should be able to debug this now.

**Comment by 0323pin** [comment]
> I'm unable to build the latest master (or even the previous 0.7.0) ...

If you want, I can do a test build on bare-metal x86_64.

**Comment by fujiapple852** [comment]
@0323pin that would be useful, thanks.  I'm going to release `0.8.0` soon so it would be good to confirm everything still works before I do.

Apart from checking that it builds from `master`, I was also hoping to run against a branch with some new logging enabled to capture more details above the failure above.  Would you be able to do that if it's not too much trouble?

```
git checkout feat-error-context
cargo build
sudo target/debug/trip example.com -m silent -v --log-span-events full > trippy.log
```

When running, resize the window and it should fail with the `Interrupted system call` error, then send me the log.  Note that the above will generate a lot of output, so i'd only run it for a few seconds before resizing the window.

**Comment by 0323pin** [comment]
> I was also hoping to run against a branch with some new logging enabled to capture more details above the failure above. Would you be able to do that if it's not too much trouble?

Yes, I can do this but, I've already built from the 5b5ca30033fee48c2389723a6641a14b21a8a1b2

```
~> trip -V
trip 0.8.0-dev
~> uname -v
NetBSD 10.99.4 (GENERIC) #0: Fri May 12 13:29:41 UTC 2023  mkrepro@mkrepro.NetBSD.org:/usr/src/sys/arch/amd64/compile/GENERIC
~> pkgin list | grep rust
rust-1.69.0          Safe, concurrent, practical language
```

![2023-05-14-130643_1366x768_scrot](https://github.com/fujiapple852/trippy/assets/90570748/328cea92-eb24-4b1f-ab42-d79d5c11b6c0)

Resizing the window causes the expected failure. I'll try to find the time to build with the logging feature enabled later today. But, at least you know it builds and runs.

**Comment by 0323pin** [comment]
Running `target/debug/trip example.com -m silent -v --log-span-events full > trippy.log` doesn't display the TUI. Is this intentional?

If so, I've e-mailed you the log-file, hopefully it comes through 19 MB.

**Comment by fujiapple852** [comment]
Thanks, I got the file.  

From the trace it doesn't look like it crashed, so I guess it only crashes when you resize when the TUI is running _and_ the backend tracing is running, which the current tracing code doesn't support.  

To fix this I think I'll need a netBSD env I can use to debug this directly (also I tried a VM in virtualBox without much luck).  What would be a good forum to ask for help for this?

Anyway good news that the latest code that it builds ok, so I can proceed with the release.

**Comment by fujiapple852** [comment]
@0323pin I was eventually able to get NetBSD running locally on macOS (intel Mac) using `qemu`.  

I downloaded:

http://nycdn.netbsd.org/pub/NetBSD-daily/netbsd-9/202308261920Z/images/NetBSD-9.3_STABLE-amd64.iso

I created the VM with:

```
qemu-img create virtualmachine.img 10G
```

For the. initial install I ran:

```
qemu-system-x86_64 -boot d -cdrom ~/Downloads/NetBSD-9.3-amd64.iso -enable-kvm -m 3G -hda virtualmachine.img
```

...and followed the prompts (picked defaults for most things, enabled sshd and nntpd)

After installation, i'm running it with (i've tried a few variation on this to try and speed things up):

```
qemu-system-x86_64 -m 3G -M q35 -cpu host -smp 4 -hda virtualmachine.img -accel hvf
```

I then installed `pkgin` (that was a struggle!).  I don't know why but pkgin runs _really_ slowly (multi-minute to search for or install a package).  I also had to change from `cdn.netbsd.org` to `ftp.netbsd.org` to get it to work at all.

I then installed `trippy` version `0.8.0`:

```
pkgin install trippy
```

It runs, though i'm obviously missing some setup as it looks like the following:

<img width="719" alt="Screenshot 2023-08-31 at 11 50 35 PM" src="https://github.com/fujiapple852/trippy/assets/36075719/c5a3b8d0-a73b-435c-87e8-87936aa07ed0">

I don't see any ICMP traffic being received, which I guess is some qemu config I need to tweak.

Because i'm working directly in the console I don't have any way to trigger the bug with a window "resize", I guess I need some kind of graphical environment?

I wasn't able to ssh into the vm from my Mac host, so i'm just working in the eqmu window that pops up.

**Comment by 0323pin** [comment]
Hi, great that you have managed to install it :)

> I don't know why but pkg_add & pkgin run really slowly (multi-minute to search for or install a package).

I've never experienced this. Ok, it's not as fast as `xbps` but, it's not slower than `apt`. Location/mirror?

> Because i'm working directly in the console I don't have any way to trigger the bug with a window "resize", I guess I need some kind of graphical environment?

There are two window managers (kind of anyway, `twm` and the default `ctwm`) in the base install, as well as Xorg and three shells.

**Comment by 0323pin** [comment]
@fujiapple852 On a second thought ...

If your intention is only to debug the window-resize issue on a disposable qemu-vm, which you do not intend to keep, NetBSD has everything you need on the base install.

Xorg is part of base and `ctwm` the default WM, so if you just run `startx` from the tty, you will be on a graphical env
A bare-bones one with a white `xterm` (create `.Xresources`, if you want to define other colors) but, nevertheless a graphic env 😄 

Sorry, if I can't give you proper `.xinitrc` and `.Xresources` files but, I haven't used modified defaults in ages. These days, I'm using `alacritty` built from git-HEAD and not `xterm` (no need for `Xresources`) and `elvish` also built from git-HEAD as my default shell, configured to start a graphical env straight from loggin.

If you, by any chance, run into issues with `.Xauthority` (not able to start the X-server) make sure your `/etc/hosts` is properly configured. It should contain the proper machine name (the name you gave your host during install) and your DNS domain and it should look like this:

```
#	$NetBSD: hosts,v 1.9 2013/11/24 07:20:01 dholland Exp $
#
# Host name database.
#
# This file contains addresses and aliases for local hosts whose names
# need to be resolvable during system boot; typically this includes only
# the address and FQDN for this machine's hostname.
#
# By default this file is consulted before DNS, so adding additional
# material here that then becomes out of date can lead to confusion.
# See nsswitch.conf(5).
#
::1			mybox.my.domain mybox
000.0.0.0		mybox.my.domain mybox
#
# RFC 1918 specifies that these networks are "internal":
# 10.0.0.0        -   10.255.255.255  (10/8 prefix)
# 172.16.0.0      -   172.31.255.255  (172.16/12 prefix)
# 192.168.0.0     -   192.168.255.255 (192.168/16 prefix)
```

Note: I've posted this on a forum a longtime ago so, I've hidden the numbers, 000.0.0.0 is actually something else. Don't touch the default, just fix domain name and hostname.

Now, you should be able to resize your terminal and reproduce the issue.

**Comment by fujiapple852** [comment]
> Are you on Matrix?

@0323pin I am now!  @fujiapple852:matrix.org

**Comment by fujiapple852** [comment]
> If your intention is only to debug the window-resize issue on a disposable qemu-vm, which you do not intend to keep, NetBSD has everything you need on the base install.

Yeh, disposable is fine.  I'd like to be able to fire up trippy on netbsd run a basic test before each release like I used to do on a cloud env.

> Xorg is part of base and ctwm the default WM, so if you just run startx from the tty, you will be on a graphical env

Wow, that just...worked :)

From the graphical env I was able to start trippy, resize the window and observe the crash.  Nice!

**Comment by c-git** [comment]
>Yeh, disposable is fine. I'd like to be able to fire up trippy on netbsd run a basic test before each release like I used to do on a cloud env.

Is there an easy way to add this to the pre-release CI?

**Comment by c-git** [comment]
I also joined matrix @one.----:matrix.org

**Comment by 0323pin** [comment]
> > Yeh, disposable is fine. I'd like to be able to fire up trippy on netbsd run a basic test before each release like I used to do on a cloud env.
> 
> Is there an easy way to add this to the pre-release CI?

I know nothing about CIs but, there's now support for NetBSD in https://cirrus-ci.com/build/6221284932583424

**Comment by fujiapple852** [comment]
@0323pin provisional fix available in https://github.com/fujiapple852/trippy/pull/670

It seems to fix it in my `qemu-vm` environment, would care to try it?

One caveat here is that I've had to temporarily downgrade `clap` as the latest version requires Rust 1.70 and the latest version available on netbsd 9.3 appears to be 1.69 (the latest rust available is 1.72).  Once netbsd bumps to 1.70 I can merge this fix, which is fine as I don't plan a trippy release for a while anyway.  If needed I _could_ back port the fix to 0.8.x and release it, but I think that may be overkill.

**Comment by fujiapple852** [comment]
> I know nothing about CIs but, there's now support for NetBSD in https://cirrus-ci.com/build/6221284932583424

That's awesome!

Now I wonder if it allows raw sockets and ICMP (GH actions do not...)

**Comment by 0323pin** [comment]
Thanks! I'll take it for a spin after the weekend. 
Sorry, family visit.

Did you see my comment on the Rust version in your branch commit?

**Comment by fujiapple852** [comment]
@0323pin yes I did, that's fine.  No rush here, we've waited this long it can wait a few more weeks :)

**Comment by 0323pin** [comment]
@fujiapple852 It works 😄 

![2023-09-02-173905_1366x768_scrot](https://github.com/fujiapple852/trippy/assets/90570748/625b9b75-da93-4d54-a603-74b9d77b4563)

Built with Rust-1.71.1, resize without crash.

**Comment by fujiapple852** [comment]
Thanks!  The fix has been Merged and will go into the `0.9.0` release.
